pkg/provisioners/gcppubsub/util/status.go

```diff
@@ -0,0 +1,117 @@
+	Secret *corev1.ObjectReference `json:"secret"`
+	// SecretKey is the key in Secret that contains the credential to use.
+	SecretKey string `json:"secretKey"`
+
+	// GCPProject is the GCP project where the Topic and Subscription exist.
+	GCPProject string `json:"gcpProject"`
+	// Topic is the name of the PubSub Topic created in GCP to represent this Channel.
+	Topic string `json:"topic,omitempty"`
+	// Subscriptions is the list of Subscriptions to this Channel and the PubSub Subscription in GCP
+	// that represents the Knative Eventing Subscription.
+	Subscriptions []GcpPubSubSubscriptionStatus `json:"subscriptions,omitempty"`
+}
+
+// GcpPubSubSubscriptionStatus represents the saved status of a gcp-pubsub Channel.
+type GcpPubSubSubscriptionStatus struct {
+	// Ref is a reference to the Knative Eventing Subscription that this status represents.
+	// +optional
+	Ref *corev1.ObjectReference `json:"ref,omitempty"`
+	// SubscriberURI is a copy of the SubscriberURI of this Subscription.
+	// +optional
+	SubscriberURI string `json:"subscriberURI,omitempty"`
+	// ReplyURI is a copy of the ReplyURI of this Subscription.
+	// +optional
+	ReplyURI string `json:"replyURI,omitempty"`
+
+	// Subscription is the name of the PubSub Subscription resource in GCP that represents this
+	// Knative Eventing Subscription.
+	Subscription string `json:"subscription,omitempty"`
+}
+
+// IsEmpty determines if this GcpPubSubChannelStatus is equivalent to &GcpPubSubChannelStatus{}. It
+// exists because slices are not compared by golang's ==.
+func (pcs *GcpPubSubChannelStatus) IsEmpty() bool {
+	if pcs.Secret != nil {
+		return false
+	}
+	if pcs.SecretKey != "" {
+		return false
+	}
+	if pcs.GCPProject != "" {
+		return false
+	}
+	if pcs.Topic != "" {
+		return false
+	}
+	if len(pcs.Subscriptions) > 0 {
+		return false
+	}
+	return true
+}
+
+// SaveRawStatus saves GcpPubSubChannelStatus to the given Channel, which should only be one whose
+// provisioner is gcp-pubsub.
+func SaveRawStatus(ctx context.Context, c *eventingv1alpha1.Channel, pcs *GcpPubSubChannelStatus) error {
+	jb, err := json.Marshal(pcs)
+	if err != nil {
+		logging.FromContext(ctx).Error("Error saving the raw status", zap.Error(err), zap.Any("pcs", pcs))
+		return err
+	}
+	c.Status.Raw = &runtime.RawExtension{
+		Raw: jb,
+	}
+	return nil
+}
+
+// ReadRawStatus reads GcpPubSubChannelStatus from the given Channel, which should only be one whose
+// provisioner is gcp-pubsub. If the raw status is not set, then the empty GcpPubSubChannelStatus is
+// returned.
+func ReadRawStatus(ctx context.Context, c *eventingv1alpha1.Channel) (*GcpPubSubChannelStatus, error) {
+	if c.Status.Raw == nil {
+		return &GcpPubSubChannelStatus{}, nil
+	}
+	bytes := c.Status.Raw.Raw
+	if len(bytes) == 0 {
+		return &GcpPubSubChannelStatus{}, nil
+	}
+	var pcs GcpPubSubChannelStatus
+	if err := json.Unmarshal(bytes, &pcs); err != nil {
+		logging.FromContext(ctx).Error("Unable to parse the raw status", zap.Error(err))
```

Comment:
This raises a general issue for me. Where should we log errors? As soon as it happens? In a single place at the 'top' (e.g. the response from the reconcile loop itself)? Both places? All intermediate functions?

I tend towards doing it in both, because I don't know the convention and want to ensure it is logged somewhere.

Reply:
This is a good question, and I think one that we should answer in contributor docs so it can be used to increase consistency for everyone's code.

I tend to agree with Peter Bourgon that error logs should be actionable and designed for humans (https://peter.bourgon.org/blog/2016/02/07/logging-v-instrumentation.html). I'm not so extreme as "never log errors that are not actionable", but I aim to consider carefully how an error message will be useful to a future human.

When a library function uses this pattern

```go
if err != nil {
  log(err)
  return err
}
```

at first glance I would guess that the library function doesn't have the context to decide whether this error is actionable or not. If not, the function should defer the decision of how to handle the error to the caller.

A notable exception is if the function has unique context that it's about to throw away. In that case, it might log an error/info or add that context to the error itself.

Reply:
Done.